Index: gcode.cpp
===================================================================
--- gcode.cpp	(révision 120)
+++ gcode.cpp	(copie de travail)
@@ -849,6 +849,18 @@
     return status;
 }
 
+/*
+/// LETARTARE  May 13, 2014
+	1- frame1 : < 8c (3 axes), 0.845 (4 axes)  	-> $$==0
+		received == "MPos:[....],WPos:[....]"
+	2- frame2 : 0.8c1 (4 axes)       			-> $$==1
+		received == "<State,MPos:...,WPos:...>"
+    3- frame2 : 0.8c (3 axes)        			-> $$==1
+		received == "<State,MPos:...,WPos:...>"
+    4- frame3 : 0.9d (3 axes), 0.9d1 (4 axes) 	-> $$==1
+		received == "<State,MPos: ... ,WPos:...,Ln=..>"
+*/
+
 void GCode::parseCoordinates(const QString& received, bool aggressive)
 {
     if (aggressive)
@@ -860,7 +872,7 @@
         parseCoordTimer.restart();
     }
 
-	bool good = false;
+	bool good = false, goodMPos = false;
 	int captureCount ;
 	QString state;
     QString prepend;
@@ -875,25 +887,38 @@
 	QString coordRegExp;
 	QRegExp rxStateMPos;
 	QRegExp rxWPos;
-	/// 3 axis
-	QString format("(-*\\d+\\.\\d+),(-*\\d+\\.\\d+)") ;
-    int maxaxis = MAX_AXIS_COUNT, naxis ;
-    for (naxis = DEFAULT_AXIS_COUNT; naxis <= maxaxis; naxis++) {
+/// 1 axis
+	QString format1("(-*\\d+\\.\\d+)") ;
+	QString sep(",");
+	/// 3 axes
+	QString format3 = format1 + sep + format1 + sep + format1 ;
+	/// 4 axes
+	QString format4 = format3 + sep + format1 ;
+	QString format ;
+    int naxis ;
+	for (naxis = MAX_AXIS_COUNT; naxis >= DEFAULT_AXIS_COUNT; naxis--) {
 		if (!doubleDollarFormat)
 			captureCount = naxis ;
 		else
 			captureCount = naxis + 1 ;
-		//
-		format += ",(-*\\d+\\.\\d+)" ;
+
+               if (naxis == MAX_AXIS_COUNT )
+			format = format4;
+		else
+			format = format3;
+
 		coordRegExp = prepend + format + append ;
 		rxStateMPos = QRegExp(preamble + coordRegExp);
-		rxWPos = QRegExp(QString("WPos:") + coordRegExp);
-		good = rxStateMPos.indexIn(received, 0) != -1
-			   && rxStateMPos.captureCount() == captureCount
-			   && rxWPos.indexIn(received, 0) != -1
-			   && rxWPos.captureCount() == naxis
-			   ;
-		// find ...
+                goodMPos = rxStateMPos.indexIn(received, 0) != -1
+					&& rxStateMPos.captureCount() == captureCount
+					;
+		if (goodMPos) {
+			rxWPos = QRegExp(QString("WPos:") + coordRegExp);
+			good = rxWPos.indexIn(received, 0) != -1
+					&& rxWPos.captureCount() == naxis
+					;
+}
+		// find naxis ...
 		if (good)
 			break;
 	}
@@ -941,15 +966,15 @@
 			workCoord.stoppedZ = false;
 
 		workCoord.sliderZIndex = sliderZCount;
+if (doubleDollarFormat)
+			diag(qPrintable(tr("Decoded: State:%s")),  qPrintable(state) );
         if (numaxis == DEFAULT_AXIS_COUNT)
-			diag(qPrintable(tr("Decoded: State:%s MPos: %f,%f,%f WPos: %f,%f,%f\n")),
-				 qPrintable(state),
+diag(qPrintable(tr("Decoded: MPos: %f,%f,%f WPos: %f,%f,%f\n")),			
 				 machineCoord.x, machineCoord.y, machineCoord.z,
 				 workCoord.x, workCoord.y, workCoord.z
 				 );
         else if (numaxis == MAX_AXIS_COUNT)
-			diag(qPrintable(tr("Decoded: State:%s MPos: %f,%f,%f,%f WPos: %f,%f,%f,%f\n")),
-				 qPrintable(state),
+diag(qPrintable(tr("Decoded: MPos: %f,%f,%f,%f WPos: %f,%f,%f,%f\n")),
                  machineCoord.x, machineCoord.y, machineCoord.z, machineCoord.fourth,
                  workCoord.x, workCoord.y, workCoord.z, workCoord.fourth
 				 );
